test/hotspot/jtreg/runtime/cds/appcds/jigsaw/module/ModuleOption.java

```diff
@@ -0,0 +1,121 @@
+        final String incubatorModule = "jdk.incubator.vector";
+        final String loggingOption = "-Xlog:cds=debug,cds+module=debug,module=trace";
+        final String versionPattern = "java.[0-9][0-9][-].*";
+        String archiveName = TestCommon.getNewArchiveName("module-option");
+        TestCommon.setCurrentArchiveName(archiveName);
+
+        // dump a base archive with -m jdk.httpserver
+        OutputAnalyzer oa = TestCommon.dumpBaseArchive(
+            archiveName,
+            loggingOption,
+            "-m", moduleOption,
+            "-version");
+        oa.shouldHaveExitValue(0);
+
+        // same module specified during runtime
+        oa = TestCommon.execCommon(
+            loggingOption,
+            "-m", moduleOption,
+            "-version");
+        oa.shouldHaveExitValue(0)
+          // version of the jdk.httpserver module, e.g. java 22-ea
+          .shouldMatch(versionPattern)
+          .shouldMatch("cds,module.*Restored from archive: entry.0x.*name jdk.httpserver");
+
+        // different module specified during runtime
+        oa = TestCommon.execCommon(
+            loggingOption,
+            "-m", "jdk.compiler/com.sun.tools.javac.Main",
+            "-version");
+        oa.shouldHaveExitValue(0)
+          .shouldContain("Mismatched modules: runtime jdk.compiler dump time jdk.httpserver");
+
+        // no module specified during runtime
+        oa = TestCommon.execCommon(
+            loggingOption,
+            "-version");
+        oa.shouldHaveExitValue(0)
+          .shouldContain("Module jdk.httpserver specified during dump time but not during runtime");
+
+        // dump an archive without the module option
+        archiveName = TestCommon.getNewArchiveName("no-module-option");
+        TestCommon.setCurrentArchiveName(archiveName);
+        oa = TestCommon.dumpBaseArchive(
+            archiveName,
+            loggingOption,
+            "-version");
+        oa.shouldHaveExitValue(0);
+
+        // run with module option
+        oa = TestCommon.execCommon(
+            loggingOption,
+            "-m", moduleOption,
+            "-version");
+        oa.shouldHaveExitValue(0)
+          .shouldContain("Module jdk.httpserver specified during runtime but not during dump time")
+          // version of the jdk.httpserver module, e.g. java 22-ea
+          .shouldMatch(versionPattern);
+
+        // dump an archive with an incubator module, -m jdk.incubator.vector
+        archiveName = TestCommon.getNewArchiveName("incubator-module");
+        TestCommon.setCurrentArchiveName(archiveName);
+        oa = TestCommon.dumpBaseArchive(
+            archiveName,
+            loggingOption,
+            "-m", incubatorModule,
+            "-version");
+        oa.shouldHaveExitValue(0)
+          // module graph won't be archived with an incubator module
+          .shouldContain("archivedBootLayer not available, disabling full module graph");
```

Comment:
For thoroughness, I think we should also check for this log in heapShared.cpp:

```
    if (record->is_full_module_graph() && !CDSConfig::is_loading_full_module_graph()) {
      if (log_is_enabled(Info, cds, heap)) {
        ResourceMark rm(THREAD);
        log_info(cds, heap)("subgraph %s cannot be used because full module graph is disabled",
                            k->external_name());
      }
      return nullptr;
    }
```

Reply:
For this case, the record is null. Per our discussion, I've added another log for the null record case:
 ```
968   if (record == nullptr) {
 969     if (log_is_enabled(Info, cds, heap)) {
 970       ResourceMark rm(THREAD);
 971       log_info(cds, heap)("subgraph %s is not recorded",
 972                           k->external_name());
 973     }
```
and updated the test accordingly.